src.ts/wallet/utxohdwallet.ts

```diff
@@ -255,100 +251,100 @@ export class UTXOHDWallet extends BaseWallet {
+                newWallet.coinType == COIN_TYPE &&
+                isUTXOAddress(address) == true);
+        }
 
-        let newWallet: UTXOHDWallet;
-        let addrIndex: number = 0;
-        let zoneIndex: number = index + 1;
+        let addrIndex: number = index;
         do {
             newWallet = this.derivePath(addrIndex.toString());
-            if (getShardForAddress(newWallet.address) == shard && ((newWallet.coinType == 969) == isUTXOAddress(newWallet.address)))
-            zoneIndex--;
             addrIndex++;
-        } while ( zoneIndex > 0);
+        } while (!isValidAddressForZone(newWallet.address));
+
+        const addresInfo = { address: newWallet.address, privKey: newWallet.privateKey, index: addrIndex - 1};
+        
+        return addresInfo;
+    }
+
 
-        return newWallet;   
+    // helper function to validate the zone
+    private validateZone(zone: string): boolean {
+        zone = zone.toLowerCase()
+        const shard = ShardData.find(shard => shard.name.toLowerCase() === zone ||
+            shard.nickname.toLowerCase() === zone ||
+            shard.byte.toLowerCase() === zone);
+        return shard !== undefined;
     }
+
+    // getOutpointsByAddress queries the network node for the outpoints of the specified address
+    private async getOutpointsByAddress(address: string): Promise<Outpoint[]> {
+        try { 
+            const outpointsMap = await this.provider!.getOutpointsByAddress(address);
+            if (!outpointsMap) {
+                return [];
+            }
+            return Object.values(outpointsMap) as Outpoint[];
+        } catch (error) {
+            throw new Error(`Failed to get outpoints for address: ${address}`);
+        }
+    }
+
     /**
-     *  Signs a UTXO transaction and returns the serialized transaction
+     * Initializes the wallet by generating addresses and private keys for the specified zone.
+     * The wallet will generate addresses until it has `GAP` number of naked addresses.
+     * A provider must be set before calling this method.
+     * @param zone - Required. Zone identifier used to validate the derived address.
+     * @returns {Promise<void>}
      */
+    public async init(zone: string): Promise<void> {
+        if (!this.validateZone(zone)) throw new Error(`Invalid zone: ${zone}`);
+        if (!this.provider) throw new Error("Provider not set");
+
+        let shardWalletData = this.#shardWalletsMap.get(zone);
+        if (!shardWalletData) {
+            shardWalletData = { addressesInfo: [], outpoints: new Map() };
+            this.#shardWalletsMap.set(zone, shardWalletData);  
+        }
+        
+        let nakedCount = 0;
+        let derivationIndex = 0;  
+    
+        while (nakedCount < GAP) {
+            const addressInfo = this.deriveAddress(derivationIndex, zone);
+            if (!addressInfo) {
+                throw new Error(`Failed to derive address with index ${derivationIndex} for zone: ${zone}`);
```

Comment:
Is this error reachable? It appears the `deriveAddress` method will take as long as needed to derive a valid address. Should we consider adding a limit to the number of attempts `deriveAddress` is allowed to make before raising an error? The limit could be something like 1000 or 2000.